url70ed135da7_ralfbiedert_jcores_tgz-pJ8-sandbox_SimpleFilterTestJ8/src/net/jcores/jre/managers/ManagerExecution.java

```diff
@@ -0,0 +1,154 @@
+ * TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE,
+ * EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ * 
+ */
+package net.jcores.jre.managers;
+
+import java.util.concurrent.BrokenBarrierException;
+import java.util.concurrent.CyclicBarrier;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ThreadFactory;
+
+import net.jcores.jre.interfaces.functions.F0;
+import net.jcores.jre.utils.internal.structures.ProfileInformation;
+
+/**
+ * Manages the execution of threads.
+ * 
+ * @since 1.0
+ * @author Ralf Biedert
+ */
+public class ManagerExecution extends Manager {
+    /** The executor with which we execute tasks. */
+    ExecutorService executor;
+    
+    /** Our profile information so we know how long new tasks took */
+    ProfileInformation profile;
+
+    /** Creates our execution manager. */
+    public ManagerExecution() {
+        // Create an executor that does not prevent us from quitting.
+        this.executor = Executors.newCachedThreadPool(new ThreadFactory() {
+            public Thread newThread(Runnable r) {
+                final Thread t = new Thread(r);
+                t.setDaemon(true);
+                return t;
+            }
+        });
+        
+        this.profile = profile();
+    }
+
+    
+    /**
+     * Returns our executor.
+     * 
+     * @return Our executor.
+     */
+    public ExecutorService getExecutor() {
+        return this.executor;
+    }
+    
+    /**
+     * Benchmark the VM. Dirty, but should give us some rough estimates
+     * 
+     * @return
+     */
+    private ProfileInformation profile() {
+        final ProfileInformation p = new ProfileInformation();
+        final int RUNS = 10;
+        final int N = 5;
+
+        // Measure how long it takes to fork a thread and to wait for it again. We
+        // test 10 times and take the average of the last 5 runs.
+        long times[] = new long[RUNS];
+        for (int i = 0; i < RUNS; i++) {
+            times[i] = measure(new F0() {
+                @Override
+                public void f() {
+                    final CyclicBarrier barrier = new CyclicBarrier(2);
+
+                    // Execute the given task in the executor.
+                    ManagerExecution.this.executor.submit(new Runnable() {
+                        @Override
+                        public void run() {
+                            try {
+                                barrier.await();
+                            } catch (InterruptedException e) {
+                                e.printStackTrace();
+                            } catch (BrokenBarrierException e) {
+                                e.printStackTrace();
+                            }
+                        }
+                    });
+
+                    try {
+                        barrier.await();
+                    } catch (InterruptedException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)